File: 2016/day23.py
# @formatter:off
from os.path import isfile, join as path_join
from sys import path as sys_path
sys_path.insert(1, path_join(sys_path[0], '..'))
file_name = path_join('input', 'day23.txt')
def to_list(mf: object = int, sep='\n'): return [mf(x) for x in open(file_name).read().split(sep) if x]
def to_gen(mf: object = int, sep='\n'): return (mf(x) for x in open(file_name).read().split(sep) if x)
def p1(*args): ans = part1(*args); print(ans); return ans
def p2(*args): ans = part2(*args); print(ans); return ans

if not isfile(file_name):
	from aoc import get_input_file
	get_input_file()
# @formatter:on

from utils import parsefile

input_text = parsefile(file_name, ["\n"])

instr = [line.split(" ") for line in input_text]

from vm import VM_v2, VM_v3


def part1():
	vm = VM_v2(instr)

	vm.registers["a"] = 7

	vm.run()
	return vm.registers["a"]


modified_input = []

loop1_2 = """cpy b c
mul a c d
cpy 0 c
cpy 0 d""".split("\n")

loop3 = """mul c 2 c
cpy 0 d""".split("\n")

modified_jnz = "jnz 1 -13".split("\n")

modified_input += input_text[0:0 + 4]
modified_input += loop1_2
modified_input += input_text[10:10 + 3]
modified_input += loop3
modified_input += input_text[16:16 + 2]
modified_input += modified_jnz
modified_input += input_text[19:19 + 7]

modified_instr = [line.split(" ") for line in modified_input]


def part2():
	vm = VM_v3(modified_instr)

	vm.registers["a"] = 12

	vm.run()
	return vm.registers["a"]


p1()
p2()

"""
part 2 loops

cpy a b
dec b
cpy a d    >4
cpy 0 a
cpy b c    >2 c=b; d=0; a+=c*d; c=0
inc a      >1 a+=c; c=0
dec c
jnz c -2   <1
dec d
jnz d -5   <2
dec b
cpy b c
cpy c d
dec d      >3 c+=d; d=0
inc c
jnz d -2   <3 
tgl c
cpy -16 c
jnz 1 c   #<4
cpy 98 c
jnz 86 d  #>6
inc a      >5
inc d     #
jnz d -2   <5
inc c     #
jnz c -5   <6

loops as functions

func 1 {
	do {
		a++
		c--
	} while c > 0
}

func 1a {
	a+=c
	c=0
}

func 2 {
	c=b
	do {
		func 1()
		d--
	} while d > 0
}

func 2a {
	c=b
	a+=c*d
	c=0
	d=0
}

func 3 {
	do {
		d--
		c++
	} while d > 0
}

func 3a {
	c+=d
	d=0
}

func 4 {
	do {
		d=a
		a=0
		func 2()
		b--
		c=b
		d=c
		func 3()
		toggle c
		if c == 2
			break
		c = -16
	} while true
}

func 4a {
	do {
		d=a
		a=0
		
		c=b
		a+=c*d
		c=0
		d=0
		
		b--
		c=b
		d=c
		
		c+=d # c=2*c
		d=0
		
		toggle c
		if c == 2
			break
		c = -16
	} while true
}

"""

File: cpp/2016/day23.cpp
#include "../aocHelper.h"

#include "vm.h"

class Day23 : public BaseDay
{
public:
	Day23() : BaseDay("23") {}

	result_type solve() override
	{
		long long part1 = 0, part2 = 0;

		VM vm{ input };

		// part 1
		vm.registers['a'] = 7;
		vm.run();
		part1 = vm.registers['a'];

		// part 2
		// modify instructions to use a multiply instead of nested loops
		vector<VM::Instruction> loop_1_2;
		// cpy b c
		loop_1_2.push_back({ VM::Opcode::CPY, {{VM::Operand::Register, 'b'}, {VM::Operand::Register, 'c'}} });
		// mul a c d
		loop_1_2.push_back({ VM::Opcode::MUL, {{VM::Operand::Register, 'a'}, {VM::Operand::Register, 'c'}, {VM::Operand::Register, 'd'}} });
		// cpy 0 c
		loop_1_2.push_back({ VM::Opcode::CPY, {{VM::Operand::Number, 0}, {VM::Operand::Register, 'c'}} });
		// cpy 0 d
		loop_1_2.push_back({ VM::Opcode::CPY, {{VM::Operand::Number, 0}, {VM::Operand::Register, 'd'}} });

		vector<VM::Instruction> loop_3;
		// mul c 2 c
		loop_3.push_back({ VM::Opcode::MUL, {{VM::Operand::Register, 'c'}, {VM::Operand::Number, 2}, {VM::Operand::Register, 'c'}} });
		// cpy 0 d
		loop_3.push_back({ VM::Opcode::CPY, {{VM::Operand::Number, 0}, {VM::Operand::Register, 'd'}} });

		// jnz 1 13
		VM::Instruction modified_jnz{ VM::Opcode::JNZ, {{VM::Operand::Number, 1}, {VM::Operand::Number, -13}} };

		vector<VM::Instruction> instructions;
		instructions.insert(instructions.end(), vm.instructions.begin(), vm.instructions.begin() + 4);
		instructions.insert(instructions.end(), loop_1_2.begin(), loop_1_2.end());
		instructions.insert(instructions.end(), vm.instructions.begin() + 10, vm.instructions.begin() + 10 + 3);
		instructions.insert(instructions.end(), loop_3.begin(), loop_3.end());
		instructions.insert(instructions.end(), vm.instructions.begin() + 16, vm.instructions.begin() + 16 + 2);
		instructions.push_back(modified_jnz);
		instructions.insert(instructions.end(), vm.instructions.begin() + 19, vm.instructions.begin() + 19 + 7);

		vm.reset();
		vm.instructions = instructions;
		vm.registers['a'] = 12;
		vm.run();
		part2 = vm.registers['a'];

		return { part1, part2 };
	}
};
